jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAggregator.java

```diff
@@ -0,0 +1,138 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>Aggregates data into a single ByteBuffer of a specified maximum size.</p>
+ * <p>The buffer automatically grows as data is written to it, up until it reaches the specified maximum size.
+ * Once the buffer is full, the aggregator will not aggregate any more bytes until its buffer is taken out,
+ * after which a new aggregate/take buffer cycle can start.</p>
+ * <p>The buffers are taken from the supplied {@link ByteBufferPool} or freshly allocated if one is not supplied.</p>
+ */
+public class ByteBufferAggregator
```

Comment:
This class seems a copy of the existing `ByteBufferAccumulator`.
I would have changed that class, but if we keep this class, then `ByteBufferAccumulator` should be deprecated and its usages removed.

Reply:
They're similar enough that `ByteBufferAccumulator` might be replaced with `ByteBufferAggregator` in all existing usages. But that should be evaluated separately from this PR IMHO.

Reply:
@gregw @sbordet I don't think this `ByteBufferAggregator` replaces the use case for the `ByteBufferAccumulator`.

The `ByteBufferAccumulator` is designed to avoid resizing copies, it uses a list of `RetainableByteBuffer`, so when it runs out of space it just allocates another buffer instead of resize+copy. At the end can copy into a single large buffer, or write the accumulated buffers without combining them.

Reply:
@lachlan-roberts Keeping a `List<RetainableByteBuffer>` would make a lot of sense if the passed in buffers could be retained, and thus no copy made.    But the passed in byte buffers are byte arrays, so they are copied once into the a `RetainableByteBuffer` which is put in the list, and then potentially copied again if a single buffer is taken (rather than the option of writing).   Note that because the pool will probably always give a bigger buffer than necessary, there will be some aggregation to fill up each buffer in the list.

The other key difference is that `ByteBufferAggregator` is used to fill a single buffer and then flush it (thus there are only initial growth copies), whilst `ByteBufferAccumulator` is used to accumulate an entire content, so there could be many growth copies.

If anything,  we could probably re-write `ByteBufferAccumulator` to internally use a `ByteBufferAggregator` to manage each buffer as it fills, using `takeRetainableByteBuffer` to take a buffer from the aggregator and to put it into the list of the accumulator.

@lachlan-roberts @lachlan-roberts @lorban what do you think?  An improvement for next release iteration?